src/drivers/bitbucket_cloud.js

```diff
@@ -22,39 +22,66 @@ class BitBucketCloud {
 
   async commentCreate(opts = {}) {
     const { projectPath } = this;
-    const { commitSha, report } = opts;
+    const { commitSha, report, update, watermark } = opts;
 
-    // Make a comment in the commit
     const commitEndpoint = `/repositories/${projectPath}/commit/${commitSha}/comments/`;
-    const commitBody = JSON.stringify({ content: { raw: report } });
-    const commitOutput = await this.request({
-      endpoint: commitEndpoint,
-      method: 'POST',
-      body: commitBody
-    });
+
+    const existingCommmit = (
+      await this.paginatedRequest({ endpoint: commitEndpoint, method: 'GET' })
+    )
+      .filter((comment) => {
+        const { content: { raw = '' } = {} } = comment;
+        return raw.endsWith(watermark);
+      })
+      .sort((first, second) => first.id < second.id)
+      .pop();
+
+    const commitOutput = (
+      await this.request({
+        endpoint:
+          commitEndpoint +
+          (update && existingCommmit ? existingCommmit.id : ''),
+        method: update && existingCommmit ? 'PUT' : 'POST',
+        body: JSON.stringify({ content: { raw: report } })
+      })
+    ).links.html.href;
 
     // Check for a corresponding PR. If it exists, also put the comment there.
-    const getPrEndpt = `/repositories/${projectPath}/commit/${commitSha}/pullrequests`;
-    const { values: prs } = await this.request({ endpoint: getPrEndpt });
+    let prs;
+    try {
+      const getPrEndpoint = `/repositories/${projectPath}/commit/${commitSha}/pullrequests`;
+      prs = await this.paginatedRequest({ endpoint: getPrEndpoint });
+    } catch (err) {
+      if (err.message === 'Not Found Resource not found')
+        console.warn(
+          "Can't create a pull request comment: the Pull Request Commit Links application has not been installed."
+        );
+      else throw err;
+    }
 
     if (prs && prs.length) {
       for (const pr of prs) {
-        try {
-          // Append a watermark to the report with a link to the commit
-          const commitLink = commitSha.substr(0, 7);
-          const longReport = `${commitLink}   \n${report}`;
-          const prBody = JSON.stringify({ content: { raw: longReport } });
-
-          // Write a comment on the PR
-          const prEndpoint = `/repositories/${projectPath}/pullrequests/${pr.id}/comments`;
-          await this.request({
-            endpoint: prEndpoint,
-            method: 'POST',
-            body: prBody
-          });
-        } catch (err) {
-          console.debug(err.message);
-        }
+        // Append a watermark to the report with a link to the commit
+        const commitLink = commitSha.substr(0, 7);
+        const longReport = `${commitLink}   \n${report}`;
```

Comment:
why the trailing WS?

```suggestion
        const longReport = `${commitLink}\n${report}`;
```

Reply:
For no specific reason, or so it seems. Unless @DavidGOrtega states otherwise, I would just remove the inner whitespace as you suggest.

Reply:
As far as I remember markdown needs a space before a new line to actually create a new line? 🤔

Reply:
did you mean this?

```suggestion
        const longReport = `${commitLink}\n\n${report}`;
```

Reply:
or this?
```js
const longReport = `${commitLink}\s\n${report}`;
```
I do not remember, I slightly remember that if you remve the space the new line wont happen

Reply:
No, he just means that Bitbucket developers... _think differently._

- `hello   \nworld` becomes `<p>hello <br>\nworld</p>`
- `hello  \nworld` becomes `<p>hello<br>\nworld</p>`
- `hello \nworld` becomes `<p>hello \nworld</p>`
- `hello\nworld` becomes `<p>hello\nworld</p>`
- `hello\n\nworld` becomes `<p>hello</p>\n<p>world</p>`
- `hello\n\n\nworld` becomes `<p>hello</p>\n<p>world</p>`

Reply:
@0x2b3bfa0, Who are you talking to? t @casperdcl? to me?

Reply:
> Bitbucket developers... think differently.

I can not agree with that more

Reply:
Sorry, https://github.com/iterative/cml/pull/581#discussion_r653814301 got stuck and I only saw https://github.com/iterative/cml/pull/581#discussion_r653810810. Now it should be clear.

Reply:
For the record the Bitbucket markdown does not behave the same and there are some tricks here and there

Reply:
so... `\n\n` is still the intention, right?

Reply:
try it and check within Bitbucket @casperdcl

Reply:
Just to spare unnecessary sufferings: that's what I did; check my comment
above.

Using `hello\n\nworld` would produce `<p>hello</p>\n<p>world</p>`

Reply:
For avoidance of doubt: I believe

- `\n\n` has and will always, in all flavours of markdown, create a new paragraph
- the intention of the code here was to create a new paragraph

so, er, what's the problem?

Reply:
## `hello␊␊world`

![Captura de pantalla 2021-06-18 a las 14 56 54](https://user-images.githubusercontent.com/11387611/122564294-78075780-d045-11eb-8570-afeccf113f3a.png)

***

## `hello␠␠␊world`
![Captura de pantalla 2021-06-18 a las 14 57 05](https://user-images.githubusercontent.com/11387611/122564295-78075780-d045-11eb-8f0e-410e8b80ce5b.png)

Reply:
> _so, er, what's the problem?_

None, I think. In fact `␊␊` produces leaves more vertical space than `␠␠␊` and, in my opinion, looks nicer.

Reply:
👍🏼  for replacing it with `\n\n`

Reply:
I would not say that two lines is nicer

Reply:
The intention is one line not  two for the record

Reply:
Ah so basically we all agree that new paragraph is the intention but:

- I don't care about the amount of vspace, but do prefer universal syntax
- David prefers less vspace by using BB markdown extension syntax
- Helio prefers more vspace by using universal markdown syntax

so in summary we either do

```suggestion
        const longReport = `${commitLink}  \n${report}`;
```

or

```suggestion
        const longReport = `${commitLink}\n\n${report}`;
```

plz pick one. toss a coin if you have to.

Reply:
It should be only one line, not two

Reply:
If Bitcoin block 688142 hash rightmost hex nibble is odd (i.e. 0xf is odd but 0xe is even) I'll agree with @DavidGOrtega. That's how we flip a coin in the 21st century! 🎲 

I would even suggest the solomonic solution of adding the commit hash as a footer instead of a header. See my next comment for a realistic example.

Reply:
## CML Report

| Value   |      Description      |
|----------|:-------------:|
| 42 |  ··· |

***

_Commit comment for 4b799ec1e7673ab716cbd2480308b71942182fdb_

Reply:
> adding the commit hash as a footer instead of a header. 

We tried that and looked worse and did not have the same feeling as their counterparts. As a header was a better option. Just my two satoshis

Reply:
Bitcoin block [688142](https://btc.com/00000000000000000007d08bdea51e4158bc2eab60e1da5413a739220adf05ab) ends in `0xb` (which is odd) so I agree with @DavidGOrtega. 🤷🏼‍♂️  😄